src/providers/twitch/TwitchChannel.cpp

```diff
@@ -1685,11 +1685,88 @@ void TwitchChannel::upsertPersonalSeventvEmotes(
         return;
     }
 
+    using MessageElementVec = std::vector<std::unique_ptr<MessageElement>>;
+
+    const auto upsertWords = [&](MessageElementVec &elements,
+                                 TextElement *textElement) {
+        std::vector<TextElement::Word> words;
+        // Append the text element and clear the vector.
+        const auto flush = [&]() {
+            elements.emplace_back(std::make_unique<TextElement>(
+                std::move(words), textElement->getFlags(), textElement->color(),
+                textElement->style()));
+            words.clear();
+        };
+
+        // Attempts to insert a zero width emote.
+        // Returns `true` if the insertion succeeded.
+        const auto tryInsertZeroWidth = [&](const EmotePtr &emote) -> bool {
+            if (!words.empty() || elements.empty())
+            {
+                // either the last element will be a TextElement _or_
+                // there are no elements.
+                return false;
+            }
+            // [THIS IS LARGELY THE SAME AS IN TwitchMessageBuilder::tryAppendEmote]
+            // Attempt to merge current zero-width emote into any previous emotes
+            auto *asEmote = dynamic_cast<EmoteElement *>(elements.back().get());
+            if (asEmote)
+            {
+                // Make sure to access asEmote before taking ownership when releasing
+                auto baseEmote = asEmote->getEmote();
+                // Need to remove EmoteElement and replace with LayeredEmoteElement
+                auto baseEmoteElement = std::move(elements.back());
+                elements.pop_back();
+
+                std::vector<EmotePtr> layers = {baseEmote, emote};
+                elements.emplace_back(std::make_unique<LayeredEmoteElement>(
+                    std::move(layers), baseEmoteElement->getFlags(),
+                    textElement->color()));
+                return true;
+            }
+
+            auto *asLayered =
+                dynamic_cast<LayeredEmoteElement *>(elements.back().get());
+            if (asLayered)
+            {
+                asLayered->addEmoteLayer(emote);
+                return true;
+            }
+            return false;
+        };
+
+        // Search for a word that matches any emote.
+        for (const auto &word : textElement->words())
+        {
+            auto emoteIt = emoteMap->find(EmoteName{word.text});
+            if (emoteIt != emoteMap->cend())
```

Comment:
Personal preference, but flipping this conditional so that the else is the only nested block might make this a bit more readable (to me) as it's the exception to the logic flow